PUBDEV-5705: Added new parameters to choose columns to skip before parsing a file

wendycwong opened the pull request:
This PR completes the work for JIRA: https://0xdata.atlassian.net/projects/PUBDEV/issues/PUBDEV-5705?filter=myopenissues.

I am close to done.....

I have added skipped_column support for:
1l. csv parser
2. arff parser
3. svmlight parser (will throw an error, does not support this feature)
4. avro parser (will throw an error, does not support this feature)
5. parquet parser
6. orc parser
7. zip parser
8. Added junit/R/Python unit tests.

Make sure the gradle.properties and setting.gradle files are not changed before merging to master or anywhere else.

Pscheidl commented on h2o-core/src/main/java/water/api/ParseHandler.java:
```diff
@@ -26,20 +22,32 @@ public ParseV3 parse(int version, ParseV3 parse) {
+            && ((setup.getSkippedColumns() != null) && (setup.getSkippedColumns().length >0)))
+      throw new H2OIllegalArgumentException("Parser: skipped_columns are not supported for SVMlight or Avro parsers.");
+
+    if ((setup.getSkippedColumns() !=null &&
```
I don't understand this condition. 

We should probably subtract `nSkipped` from `nParsed`, where `nSkipped` is number of skipped cols and `nParsed` is number of columns parsed from the dataset. If the result is 0 (assert for it never to be negative ?), then the exception is thrown ? Plus the null checks of course.

Am I missing something ?

Pscheidl commented on h2o-core/src/main/java/water/parser/CsvParser.java:
```diff
@@ -59,7 +58,8 @@
     int tokenStart = 0; // used for numeric token to backtrace if not successful
-    int colIdx = 0;
+    int columnCounter = 0;  // index into parsed columns only, exclude skipped columns.
+    int colIdx = 0; // count each actual column in the dataset including the skipped columns
```
This should not be named `colIdx` but rather something like `colCnt` ? A minor thing, but when looking at the variable without the comment, it makes me think about a stored ID of an actual column. Is that the purpose ?

wendycwong replied:
This is legacy from Tomas (or whoever implement it).  It is column index in the short form.  It indexes into each column including those to be skipped.

Pscheidl commented on h2o-core/src/main/java/water/parser/CsvParser.java:
```diff
@@ -80,10 +80,17 @@
     dout.newLine();
 
     final boolean forceable = dout instanceof FVecParseWriter && ((FVecParseWriter)dout)._ctypes != null && _setup._column_types != null;
+/*    if (_setup.getSkippedColumns() !=null && _setup.get_parse_columns_indices()==null)
```
Remove commented code ?

wendycwong replied:
Good idea.  Done

Pscheidl commented:
I wonder if method `protected int fileHasHeader(byte[] bits, ParseSetup ps)` in `CsvParser` is going to work with columns skipped as well. Probably yes, as What if the user provides custom header names only for a subset of columns and relies on column skipping ?

wendycwong commented on gradle.properties:
```diff
@@ -9,7 +9,7 @@ doFindbugs=false
 
 # The flag to include ORC support inside default h2o.jar.
-doIncludeOrc=false
+doIncludeOrc=true
```
Good catch pavel.  I will fix it after I finished all parsers.  Needed it to test out orc parser.

wendycwong commented on h2o-core/src/main/java/water/api/ParseHandler.java:
```diff
@@ -26,20 +22,32 @@ public ParseV3 parse(int version, ParseV3 parse) {
     for (int i = 0; i < parse.source_frames.length; i++)
       srcs[i] = parse.source_frames[i].key();
 
+    if ((setup.getParseType().name().toLowerCase().equals("svmlight") ||
```
I am too lazy to figure out exactly what the correct type is.  I found sometimes it says Svmlight or SVMLIGHT.  Doing it this way, I will catch everything unless of course I mis-spelled the name.

wendycwong replied:
We do not have a parseInfo for each parser type.  So, this cannot be done.  Sorry.

wendycwong commented on h2o-core/src/main/java/water/api/ParseHandler.java:
```diff
@@ -26,20 +22,32 @@ public ParseV3 parse(int version, ParseV3 parse) {
       srcs[i] = parse.source_frames[i].key();
 
+    if ((setup.getParseType().name().toLowerCase().equals("svmlight") ||
+            (setup.getParseType().name().toLowerCase().equals("avro") ))
```
That is a good idea.  Yes, I believe so.  Let me find it and fix it.

wendycwong replied:
Nope, I checked.  OrcParser has OrcParseInfo but not sure about the other ones.  We only go through the parseHandler a few times before the parse.  It should not cost too much time.

I am afraid that people like to change parser names, if they do, this will break.....

wendycwong commented on h2o-core/src/main/java/water/parser/FVecParseWriter.java:
```diff
@@ -27,15 +26,51 @@
+  int[] _parse_columns_indices;
+
+  // note that if parse_columns_indices==null, it imples all columns are parsed.
+  public FVecParseWriter(Vec.VectorGroup vg, int cidx, Categorical[] categoricals, byte[] ctypes, int chunkSize,
```
Michal:  

This invocation is only called in water.parser, ParseDataset.java and no where else.  So backwards compatibility is not an issue here at the moment.

michalkurka commented on h2o-core/src/test/java/water/parser/ParseFolderTest.java:
```diff
@@ -1,16 +1,19 @@
+import java.io.File;
+
 public class ParseFolderTest extends TestUtil {
-  @BeforeClass static public void setup() { stall_till_cloudsize(5); }
```
Please fix

wendycwong replied:
Done

michalkurka commented on h2o-parsers/h2o-parquet-parser/src/main/java/water/parser/parquet/ChunkConverter.java:
```diff
@@ -69,6 +71,106 @@ long getCurrentRecordIdx() {
+
+
+  private static class NullStringConverter extends PrimitiveConverter {
+    private final boolean _dictionarySupport;
```
please move all of these to NullConverter

wendycwong replied:
Done

wendycwong commented:
Tips for code review:

To enable skipping columns during parse, introduced two parameters to ParseSetup.java
1. _skipped_columns (int[] with inputs from user listing column indices to skip)
2. _parse_columns_indices (int[] representing columns indices to be included in the final frame).  Size of this array is number of columns in file minus number of skipped columns.

Added one parameter to parser.java
boolean[] _keepColumns with size = number of columns in file.  For skipped columns indices, its value is false.

-----------------------------------------------------------------
During guessSetup or when users did not specify skipped_columns, _parse_columns_indices will contain all column indices and _keepColumns will contain all True.  Remember that the various parsers associated with the different parsers will be called once to attempt to parse the file with that parser type.  Hence, whatever change I made MUST NOT affect this process.  By setting up the _parse_columns_indices and _keepColumns the way I did will guarantee this process will run smoothly.

---------------------------------------------------------------
I have not changed anything associate with reading the input file because all the columns must be read even though not all of them will be included in the final parse file.

The major change is in method makedout in ParseDataset.java.  Inside method FVecParseWriter of FVecParseWriter.java, the following parameters will have dimension equals to the parse_column_indices: _ctypes, _categoricals, _vec, _nvs.

There are special cases to setting them when they are called from gzip files.

-------------------------------------------------------------
The stage is now set for parseChunk method.  Basically, we use two indices to refer to the columns.  One is to index through all the columns in the input file, the other one refers only to the columns that are to be written to dout.

After parsing and during reconciliation of categorical levels, the same thing apply.  We only look at column indices that are inside parse_column_indices even though the Categorical variable contains all the column indices.  Only the columns that are being parsed will have non-empty hashmap representing the categorical value and integer levels.

Hope this helps, Wendy

-----------------------------------------------------------------

wendycwong commented:
This PR was broken down into 4 separate PRs and were merged.
